Tests/UnitTests/Core/Seeding/BinnedGroupTest.cpp

```diff
@@ -0,0 +1,256 @@
+  const Acts::BinnedGroup<grid_2d_t> group_2d_moved(std::move(group_2d));
+  const Acts::BinnedGroup<grid_3d_t> group_3d_moved = std::move(group_3d);
+
+  [[maybe_unused]] const grid_1d_t& retrievedGrid = group_1d.grid();
+}
+
+BOOST_AUTO_TEST_CASE(binned_group_iterations_1d_emptyGrid) {
+  using grid_t =
+      Acts::Grid<std::vector<std::size_t>, Acts::detail::EquidistantAxis>;
+  using binfinder_t = Acts::GridBinFinder<1ul>;
+
+  Acts::detail::EquidistantAxis xAxis(0, 100, 10);
+  grid_t grid(std::make_tuple(std::move(xAxis)));
+  std::unique_ptr<binfinder_t> binfinder = std::make_unique<binfinder_t>(0);
+  Acts::BinnedGroup<grid_t> group(std::move(grid), *binfinder.get(),
+                                  *binfinder.get());
+
+  Acts::BinnedGroupIterator<grid_t> itrStart = group.begin();
+  Acts::BinnedGroupIterator<grid_t> itrStop = group.end();
+
+  std::size_t nIterations = 0ul;
+  for (; itrStart != itrStop; ++itrStart, ++nIterations) {
+    [[maybe_unused]] auto candidates = *itrStart;
+  }
+
+  BOOST_CHECK_EQUAL(nIterations, 0ul);
+}
+
+BOOST_AUTO_TEST_CASE(binned_group_iterations_2d_emptyGrid) {
+  using grid_t =
+      Acts::Grid<std::vector<std::size_t>, Acts::detail::EquidistantAxis,
+                 Acts::detail::EquidistantAxis>;
+  using binfinder_t = Acts::GridBinFinder<2ul>;
+
+  Acts::detail::EquidistantAxis xAxis(0, 100, 10);
+  Acts::detail::EquidistantAxis yAxis(0, 100, 10);
+  grid_t grid(std::make_tuple(std::move(xAxis), std::move(yAxis)));
+  std::unique_ptr<binfinder_t> binfinder = std::make_unique<binfinder_t>(0, 0);
+  Acts::BinnedGroup<grid_t> group(std::move(grid), *binfinder.get(),
+                                  *binfinder.get());
+
+  Acts::BinnedGroupIterator<grid_t> itrStart = group.begin();
+  Acts::BinnedGroupIterator<grid_t> itrStop = group.end();
+
+  std::size_t nIterations = 0ul;
+  for (; itrStart != itrStop; ++itrStart, ++nIterations) {
+    [[maybe_unused]] auto candidates = *itrStart;
+  }
+
+  BOOST_CHECK_EQUAL(nIterations, 0ul);
+}
+
+BOOST_AUTO_TEST_CASE(binned_group_iterations_1d_perFilledGrid) {
+  using grid_t =
+      Acts::Grid<std::vector<std::size_t>, Acts::detail::EquidistantAxis>;
+  using binfinder_t = Acts::GridBinFinder<1ul>;
+
+  Acts::detail::EquidistantAxis xAxis(0, 100, 10);
+  grid_t grid(std::make_tuple(xAxis));
+  /// Add some entries to the grid
+  grid.at(1ul).push_back(4ul);
+  grid.at(1ul).push_back(1ul);
+  grid.at(8ul).push_back(7ul);
+  grid.at(9ul).push_back(2ul);
+
+  std::unique_ptr<binfinder_t> botBinfinder = std::make_unique<binfinder_t>(0);
+  std::unique_ptr<binfinder_t> topBinfinder = std::make_unique<binfinder_t>(1);
+  Acts::BinnedGroup<grid_t> group(std::move(grid), *botBinfinder.get(),
+                                  *topBinfinder.get());
+
+  std::size_t nIterations = 0ul;
+  for (const auto [bottom, middle, top] : group) {
+    ++nIterations;
+    BOOST_CHECK_EQUAL(bottom.size(), 1ul);
+    BOOST_CHECK_EQUAL(top.size(), 3ul);
```

Comment:
Sorry, I'm slow: how do these numbers come to be? Why is it the same for all iterations?
And where does the `nIterations == 3` come from?

Reply:
The binned group iterator will only iterate on the bins that are not empty. This is to avoid calling the `operator*` which is expensive (it handles the identification of neighbouring bins).

In this test, I'm only filling 3 bins with some values. Thus `nIterations == 3`.

Then, the bottom and top sizes are given by the `GridBinFinder`s we defined above. 

Imagine you are in the bin `(N)`, the top bin finder is requiring to find the neighbours within 1 grid bins. Thus it will return `N-1`, `N` and `N+1`. `Thus top.size() == 3`.

The bottom bin finder is requiring to find the neighbours within 0 grid bins. So it will only return `N`. As such `bottom.size() == 3`

Reply:
you can also do more complicated bin searches by providing pairs of ints btw

For instance, if you provide `{k, l}` to the bin finders, then you are instructing the code to fine the neighbouring bins  in the range `N-k, ... N .... N+l`

And you can do it for every axis independently. As well as providing different `k` and `l` for each bin in the same axis. For instance a set of pairs like this will give you different bottom and top sizes depending of which `N` bin you are: `{0, 1}, {1, 1}, {1, 0}, {2, 5}`